### ISSUE TITLE
should add exemptions (פטור ממכרז) to pipelines (based on the legacy tenders/exemption_scraper)

### ISSUE THREAD

**Issue by akariv** [OpenBudget/budgetkey-data-pipelines#4]
Exemption scraper

Exemptions are records for planned purchases of the government that are exempt from a tender process.

The website to scrape is quite slow and buggy, so we want to scrape frequently, and only get the most recent records (for example, scrape every few hours and get a day's worth of records).
(For initialisation there should also be a one-time-run pipeline that scrapes the entire website)

Data is written to a DB table (exemptions), with the `update` mode - so that records with the same publication id will be updated and data is appended and not replaced.

Scraped data includes the basic list of fields for each exemption record, as well as the list of documents for the record (which should be saved as a JSON object).

There might be issues regarding scraping from non Israeli IPs - talk to me if you encounter such issues.

---- 
Ref: 
Legacy scraper folder: https://www.dropbox.com/sh/xan213ilqrdh43p/AAC1tSoTiH1rnMOue7pBJyhHa?dl=0
Legacy scraper is in tenders/exemption_scraper.py

**Comment by OriHoch** [comment]
looks good, thanks, I'll start working on it

**Comment by OriHoch** [comment]
### status update `April 30, 2017`
after PR #8 is merged the following pipelines will be available
* `./procurement/tenders/exemptions/publishers` - gets a list of all publisher ids and names
* `./procurement/tenders/exemptions/publisher-urls` - gets a list of exemption urls for each publisher id
##### TODO
* add a pipeline for getting the publisher urls and extracting each exemption's data
  * this is mostly just the code from the legacy scraper `exemption_extended_data_web_page.extract_page_data` method
* figure out how to handle errors / timeouts / unexpected problems
  * currently it just runs over all the publisher urls and if any one fails it drops all
  * need to have some kind of retry mechanism, or other error handling
* figure out how it will run for the first time - to get the baseline and how it will run periodically - where it is configured etc..

**Comment by akariv** [comment]
Re your *TODO*:
* Timeouts / errors - assuming it's statistical (and not frequent), simply retry all network requests until successful (with a minute of delay after a failure). If 10 unsuccessful attempts - fail the processor.
* Running the 1st time / periodically: You can follow a similar scheme to what's being done in the company details and guidestar scrapers:
  - We join the current list of records with the new list of ids (based on id)
  - We push that to the scraper, which only scrapes if data is missing (or needs updating).
That pipeline runs every X hours. When the DB is empty, it will scrape everything. When it's not, it will scrape only missing records.

**Comment by OriHoch** [comment]
* added retry logic in #8 

of course, forgot an important **TODO** - add a processor that updates the DB

**Comment by OriHoch** [comment]
issue is complete once PR #13 is merged

changes from the original description / comments:
* the documents are updated in a related one to many DB table
* updating is done by limiting downloading to 5 pages (*10 exemptions per page) every run - this is very simplistic and might cause problems (or maybe not..) - see follow-up issue #34 
* running for first time can be done by setting environment variable which disables the pages limit (see the PR for details)

**Comment by OriHoch** [comment]
@akariv please review, let me know if more work is needed

(I prefer to move on to another task, but if you think something is critical for fixing let me know..)
